<!DOCTYPE html><html lang="zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0,viewport-fit=cover"><title>集群问题 | Hexo</title><meta name="author" content="CJ"><meta name="copyright" content="CJ"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="问题一 现象：Grafana触发了报警，但是没有发送报警信息，journalctl -u grafana -n 20 查看日志如下1211月 01 13:54:01 cos-bigdata-mysql grafana-server[21140]: t&#x3D;2021-11-01T13:54:01+0800 lvl&#x3D;info msg&#x3D;Rendering logger&#x3D;rendering path&#x3D;&amp;quo">
<meta property="og:type" content="article">
<meta property="og:title" content="集群问题">
<meta property="og:url" content="http://example.com/2023/05/06/%E5%A4%A7%E6%95%B0%E6%8D%AE%E7%A6%BB%E7%BA%BF/%E9%9B%86%E7%BE%A4%E9%97%AE%E9%A2%98/index.html">
<meta property="og:site_name" content="Hexo">
<meta property="og:description" content="问题一 现象：Grafana触发了报警，但是没有发送报警信息，journalctl -u grafana -n 20 查看日志如下1211月 01 13:54:01 cos-bigdata-mysql grafana-server[21140]: t&#x3D;2021-11-01T13:54:01+0800 lvl&#x3D;info msg&#x3D;Rendering logger&#x3D;rendering path&#x3D;&amp;quo">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://i.loli.net/2021/02/24/5O1day2nriDzjSu.png">
<meta property="article:published_time" content="2023-05-06T05:31:21.063Z">
<meta property="article:modified_time" content="2023-05-06T05:31:21.063Z">
<meta property="article:author" content="CJ">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://i.loli.net/2021/02/24/5O1day2nriDzjSu.png"><link rel="shortcut icon" href="/img/favicon.png"><link rel="canonical" href="http://example.com/2023/05/06/%E5%A4%A7%E6%95%B0%E6%8D%AE%E7%A6%BB%E7%BA%BF/%E9%9B%86%E7%BE%A4%E9%97%AE%E9%A2%98/index.html"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox/fancybox.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false},
  copy: {
    success: '复制成功',
    error: '复制错误',
    noSupport: '浏览器不支持'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  dateSuffix: {
    just: '刚刚',
    min: '分钟前',
    hour: '小时前',
    day: '天前',
    month: '个月前'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isAnchor: false,
  percent: {
    toc: true,
    rightside: false,
  }
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: '集群问题',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2023-05-06 13:31:21'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
    win.getCSS = (url,id = false) => new Promise((resolve, reject) => {
      const link = document.createElement('link')
      link.rel = 'stylesheet'
      link.href = url
      if (id) link.id = id
      link.onerror = reject
      link.onload = link.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        link.onload = link.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(link)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
    const detectApple = () => {
      if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
    })(window)</script><meta name="generator" content="Hexo 6.3.0"></head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src="https://i.loli.net/2021/02/24/5O1day2nriDzjSu.png" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="sidebar-site-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">419</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">0</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">38</div></a></div><hr/></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header"><nav id="nav"><span id="blog-info"><a href="/" title="Hexo"><span class="site-name">Hexo</span></a></span><div id="menus"><div id="toggle-menu"><a class="site-page" href="javascript:void(0);"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">集群问题</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">发表于</span><time class="post-meta-date-created" datetime="2023-05-06T05:31:21.063Z" title="发表于 2023-05-06 13:31:21">2023-05-06</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">更新于</span><time class="post-meta-date-updated" datetime="2023-05-06T05:31:21.063Z" title="更新于 2023-05-06 13:31:21">2023-05-06</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E5%A4%A7%E6%95%B0%E6%8D%AE%E7%A6%BB%E7%BA%BF/">大数据离线</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title="集群问题"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">阅读量:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><h3 id="问题一"><a href="#问题一" class="headerlink" title="问题一"></a>问题一</h3><ul>
<li>现象：Grafana触发了报警，但是没有发送报警信息，journalctl -u grafana -n 20 查看日志如下<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">11月 01 13:54:01 cos-bigdata-mysql grafana-server[21140]: t=2021-11-01T13:54:01+0800 lvl=info msg=Rendering logger=rendering path=&quot;d-solo/Jui9KoK7z/bigdata-server-process-alert?orgId=1&amp;panelId=2&quot;</span><br><span class="line">11月 01 13:54:01 cos-bigdata-mysql grafana-server[21140]: t=2021-11-01T13:54:01+0800 lvl=eror msg=&quot;Alert Panic&quot; logger=alerting.engine error=&quot;runtime error: invalid memory address or nil pointer dereference&quot; stack=&quot;[engine.go:181 panic.go:1038 panic.go:221 signal_unix.go:735 rendering.go:216 rendering.go:178 notifier.go:218 notifier.go:117 result_handler.go:102 engine.go:224]&quot;</span><br></pre></td></tr></table></figure></li>
<li>原因：不知道什么原因，网上查询不到。</li>
<li>解决：重启grafana即可</li>
</ul>
<h3 id="问题二"><a href="#问题二" class="headerlink" title="问题二"></a>问题二</h3><ul>
<li>现象：Grafana Alert推送告警图片失败，日志如下<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">11月 01 14:08:01 cos-bigdata-mysql grafana-server[29048]: t=2021-11-01T14:08:01+0800 lvl=info msg=Rendering logger=rendering renderer=plugin path=&quot;d-solo/Jui9KoK7z/bigdata-server-process-alert?orgId=1&amp;panelId=2&quot;</span><br><span class="line">11月 01 14:08:01 cos-bigdata-mysql grafana-server[29048]: t=2021-11-01T14:08:01+0800 lvl=eror msg=&quot;Render request failed&quot; logger=plugins.backend pluginId=grafana-image-renderer url=&quot;http://localhost:3000/d-solo/Jui9KoK7z/bigdata-server-process-alert?orgId=1&amp;panelId=2&amp;render=1&quot; error=&quot;Error: Failed to launch the browser process!</span><br><span class="line">/opt/module/grafana-8.2.2/data/plugins/grafana-image-renderer/chrome-linux/chrome: error while loading shared libraries: libatk-1.0.so.0: cannot open shared object file: No such file or directory</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">TROUBLESHOOTING: https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md</span><br><span class="line">&quot;</span><br><span class="line">11月 01 14:08:01 cos-bigdata-mysql grafana-server[29048]: t=2021-11-01T14:08:01+0800 lvl=eror msg=&quot;Failed to render and upload alert panel image.&quot; logger=alerting.notifier ruleId=2 error=&quot;rendering failed: Error: Failed to launch the browser process!</span><br><span class="line">/opt/module/grafana-8.2.2/data/plugins/grafana-image-renderer/chrome-linux/chrome: error while loading shared libraries: libatk-1.0.so.0: cannot open shared object file: No such file or directory</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">TROUBLESHOOTING: https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md</span><br><span class="line">&quot;</span><br></pre></td></tr></table></figure></li>
<li>原因：环境缺少部分依赖包，需要进行安装</li>
<li>解决：通过命令<code>ldd grafana-8.2.2/data/plugins/grafana-image-renderer/chrome-linux/chrome</code>查看缺失的环境依赖并安装。<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">yum install -y libatk-1.0.so.0 libatk-bridge-2.0.so.0 libcups.so.2 libxcb.so.1 libxkbcommon.so.0 libX11.so.6 libXcomposite.so.1 libXdamage.so.1 libXext.so.6 libXfixes.so.3 libXrandr.so.2 libgbm.so.1 libpango-1.0.so.0 libcairo.so.2 libatspi.so.0 libxshmfence.so.1</span><br></pre></td></tr></table></figure>
重新安装插件 grafana-image-renderer。<br>又出现了新问题<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">11月 01 14:45:00 cos-bigdata-mysql grafana-server[30508]: t=2021-11-01T14:45:00+0800 lvl=info msg=Rendering logger=rendering renderer=plugin path=&quot;d-solo/Jui9KoK7z/bigdata-server-process-alert?orgId=1&amp;from=1635747295710&amp;to=1635749095710&amp;panelId=2&amp;width=1000&amp;height=500&amp;tz=Asia%2FShanghai&quot;</span><br><span class="line">11月 01 14:45:00 cos-bigdata-mysql grafana-server[30508]: t=2021-11-01T14:45:00+0800 lvl=eror msg=&quot;Rendering failed.&quot; logger=context userId=1 orgId=1 uname=admin error=&quot;rpc error: code = Canceled desc = grpc: the client connection is closing&quot;</span><br><span class="line">11月 01 14:45:00 cos-bigdata-mysql grafana-server[30508]: t=2021-11-01T14:45:00+0800 lvl=eror msg=&quot;Request Completed&quot; logger=context userId=1 orgId=1 uname=admin method=GET path=/render/d-solo/Jui9KoK7z/bigdata-server-process-alert status=500 remote_addr=192.168.32.56 time_ms=5 size=1344 referer=</span><br></pre></td></tr></table></figure>
部署完成之后运行任务，又出来一个错<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">17-11-2021 15:07:18 CST customer INFO - com.alibaba.datax.common.exception.DataXException: Code:[Framework-03], Description:[DataX引擎配置错误，该问题通常是由于DataX安装错误引起，请联系您的运维解决 .].  - 在有总bps限速条件下，单个channel的bps值不能为空，也不能为非正数</span><br></pre></td></tr></table></figure></li>
</ul>
<h3 id="问题三"><a href="#问题三" class="headerlink" title="问题三"></a>问题三</h3><ul>
<li>现象：配置都一样，但是179节点的tez执行任务时报错如下<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">com.google.protobuf.ServiceException: java.io.IOException: DestHost:destPort bigdata3:43799 , LocalHost:localPort bigdata1/192.168.101.179:0. Failed on local exception: java.io.IOException: Connection reset by peer</span><br><span class="line">Caused by: java.io.IOException: DestHost:destPort bigdata3:43799 , LocalHost:localPort bigdata1/192.168.101.179:0. Failed on local exception: java.io.IOException: Connection reset by peer</span><br><span class="line">Caused by: java.io.IOException: Connection reset by peer</span><br><span class="line">2022-04-19T16:43:25,448  INFO [f8b8e0df-ccd6-4f08-ab69-225144b91dc6 main] client.TezClient: Failed to retrieve AM Status via proxy</span><br><span class="line">com.google.protobuf.ServiceException: java.io.EOFException: End of File Exception between local host is: &quot;bigdata1/192.168.101.179&quot;; destination host is: &quot;bigdata2&quot;:38361; : java.io.EOFException; For more details see:  http://wiki.apache.org/hadoop/EOFException</span><br><span class="line">Caused by: java.io.EOFException: End of File Exception between local host is: &quot;bigdata1/192.168.101.179&quot;; destination host is: &quot;bigdata2&quot;:38361; : java.io.EOFException; For more details see:  http://wiki.apache.org/hadoop/EOFException</span><br><span class="line">Caused by: java.io.EOFException]</span><br></pre></td></tr></table></figure></li>
<li>原因：以上信息不准，需要在历史服务器上看具体的异常，可以发现缺少jar包导致<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">java.lang.NoClassDefFoundError: com/fasterxml/jackson/core/exc/InputCoercionException</span><br><span class="line">Caused by: java.lang.ClassNotFoundException: com.fasterxml.jackson.core.exc.InputCoercionException</span><br></pre></td></tr></table></figure></li>
<li>解决：hive-env.sh文件中加载了Tez-0.10.1目录下的所有jar包，只需要将缺失的jackson-core-2.7.8.jarr和jackson-core-asl-1.9.13.jar放到Tez-0.10.1目录下即可。</li>
</ul>
<h3 id="问题四"><a href="#问题四" class="headerlink" title="问题四"></a>问题四</h3><ul>
<li>现象：在执行任务中Datanode掉线，重启后报错如下<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">2022-06-06 00:40:43,940 ERROR org.apache.hadoop.hdfs.server.datanode.DataNode: bigdata1:9866:DataXceiver error processing WRITE_BLOCK operation  src: /192.168.101.180:50192 dst: /192.168.101.179:9866</span><br><span class="line">java.io.IOException: Not ready to serve the block pool, BP-863358393-192.168.101.179-1626835277985.</span><br><span class="line">	at org.apache.hadoop.hdfs.server.datanode.DataXceiver.checkAndWaitForBP(DataXceiver.java:1412)</span><br><span class="line">	at org.apache.hadoop.hdfs.server.datanode.DataXceiver.checkAccess(DataXceiver.java:1428)</span><br><span class="line">	at org.apache.hadoop.hdfs.server.datanode.DataXceiver.writeBlock(DataXceiver.java:717)</span><br><span class="line">	at org.apache.hadoop.hdfs.protocol.datatransfer.Receiver.opWriteBlock(Receiver.java:173)</span><br><span class="line">	at org.apache.hadoop.hdfs.protocol.datatransfer.Receiver.processOp(Receiver.java:107)</span><br><span class="line">	at org.apache.hadoop.hdfs.server.datanode.DataXceiver.run(DataXceiver.java:292)</span><br><span class="line">	at java.lang.Thread.run(Thread.java:748)</span><br></pre></td></tr></table></figure></li>
<li>原因：字面理解为文件操作超租期，实际上就是data stream操作过程中文件被删掉了。之前也遇到过，通常是因为Mapred多个task操作同一个文件，一个task完成后删掉文件导致。</li>
<li>解决：过一会自己就好了。网上查询到的解决方法为<br> 1、datanode所在的linux服务器提高文件句柄参数；<br> 2、增加<a target="_blank" rel="noopener" href="https://so.csdn.net/so/search?q=HDFS&spm=1001.2101.3001.7020">HDFS</a>的datanode句柄参数：dfs.datanode.max.transfer.threads<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">&lt;property&gt;</span><br><span class="line">  &lt;name&gt;dfs.datanode.max.transfer.threads&lt;/name&gt;</span><br><span class="line">  &lt;value&gt;8192&lt;/value&gt;</span><br><span class="line">&lt;/property&gt;</span><br></pre></td></tr></table></figure></li>
</ul>
<h3 id="踩坑十六"><a href="#踩坑十六" class="headerlink" title="踩坑十六"></a>踩坑十六</h3><ul>
<li>现象：使用flume1.7.0向hadoop3.1.3导入数据，报如下错误<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">Exception in thread &quot;SinkRunner-PollingRunner-DefaultSinkProcessor&quot; java.lang.NoSuchMethodError: com.google.common.base.Preconditions.checkArgument(ZLjava/lang/String;Ljava/lang/Object;)V</span><br><span class="line">	at org.apache.hadoop.conf.Configuration.set(Configuration.java:1357)</span><br><span class="line">	at org.apache.hadoop.conf.Configuration.set(Configuration.java:1338)</span><br><span class="line">	at org.apache.hadoop.conf.Configuration.setBoolean(Configuration.java:1679)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter.open(BucketWriter.java:206)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter.append(BucketWriter.java:504)</span><br><span class="line">	at org.apache.flume.sink.hdfs.HDFSEventSink.process(HDFSEventSink.java:406)</span><br><span class="line">	at org.apache.flume.sink.DefaultSinkProcessor.process(DefaultSinkProcessor.java:67)</span><br><span class="line">	at org.apache.flume.SinkRunner$PollingRunner.run(SinkRunner.java:145)</span><br><span class="line">	at java.lang.Thread.run(Thread.java:748)</span><br></pre></td></tr></table></figure></li>
<li>原因：${HADOOP_HOME}&#x2F;share&#x2F;hadoop&#x2F;common&#x2F;lib下的guava包版本为guava-27.0-jre.jar，而flume-1.7.0使用的是低版本guava。版本不一致导致的错误。</li>
<li>解决：复制guava-27.0-jre.jar替换flume中低版本的guava包。</li>
</ul>
<h3 id="踩坑四十八"><a href="#踩坑四十八" class="headerlink" title="踩坑四十八"></a>踩坑四十八</h3><ul>
<li>现象：Flume同步全量数据是报错如下<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br></pre></td><td class="code"><pre><span class="line">22 三月 2022 21:12:47,055 INFO  [hdfs-k1-roll-timer-0] (org.apache.flume.sink.hdfs.BucketWriter.close:357)  - Closing /origin_data/compass/edb/edb_order/1970-01-01/edb.1647953566601.lzo.tmp</span><br><span class="line">22 三月 2022 21:13:00,650 WARN  [hdfs-k1-roll-timer-0] (org.apache.flume.sink.hdfs.BucketWriter.close:364)  - failed to close() HDFSWriter for file (/origin_data/compass/edb/edb_order/1970-01-01/edb.1647953566601.lzo.tmp). Exception follows.</span><br><span class="line">java.io.IOException: Callable timed out after 10000 ms on file: /origin_data/compass/edb/edb_order/1970-01-01/edb.1647953566601.lzo.tmp</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter.callWithTimeout(BucketWriter.java:682)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter.close(BucketWriter.java:361)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter$2.call(BucketWriter.java:275)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter$2.call(BucketWriter.java:269)</span><br><span class="line">	at java.util.concurrent.FutureTask.run(FutureTask.java:266)</span><br><span class="line">	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)</span><br><span class="line">	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)</span><br><span class="line">	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)</span><br><span class="line">	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)</span><br><span class="line">	at java.lang.Thread.run(Thread.java:748)</span><br><span class="line">Caused by: java.util.concurrent.TimeoutException</span><br><span class="line">	at java.util.concurrent.FutureTask.get(FutureTask.java:205)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter.callWithTimeout(BucketWriter.java:675)</span><br><span class="line">	... 9 more</span><br><span class="line">22 三月 2022 21:13:16,423 WARN  [hdfs-k1-roll-timer-0] (org.apache.flume.sink.hdfs.BucketWriter.close:390)  - failed to rename() file (/origin_data/compass/edb/edb_order/1970-01-01/edb.1647953566601.lzo.tmp). Exception follows.</span><br><span class="line">java.io.IOException: Callable timed out after 10000 ms on file: /origin_data/compass/edb/edb_order/1970-01-01/edb.1647953566601.lzo.tmp</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter.callWithTimeout(BucketWriter.java:682)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter.renameBucket(BucketWriter.java:614)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter.close(BucketWriter.java:388)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter$2.call(BucketWriter.java:275)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter$2.call(BucketWriter.java:269)</span><br><span class="line">	at java.util.concurrent.FutureTask.run(FutureTask.java:266)</span><br><span class="line">	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)</span><br><span class="line">	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)</span><br><span class="line">	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)</span><br><span class="line">	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)</span><br><span class="line">	at java.lang.Thread.run(Thread.java:748)</span><br><span class="line">Caused by: java.util.concurrent.TimeoutException</span><br><span class="line">	at java.util.concurrent.FutureTask.get(FutureTask.java:205)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter.callWithTimeout(BucketWriter.java:675)</span><br><span class="line">	... 10 more</span><br><span class="line">22 三月 2022 21:13:23,577 INFO  [hdfs-k1-roll-timer-0] (org.apache.flume.sink.hdfs.HDFSEventSink$1.run:382)  - Writer callback called.</span><br><span class="line">22 三月 2022 21:16:33,706 WARN  [hdfs-k1-roll-timer-0] (org.apache.flume.sink.hdfs.BucketWriter$4.call:332)  - Renaming file: /origin_data/compass/edb/edb_order/1970-01-01/edb.1647953566601.lzo.tmp failed. Will retry again in 180 seconds.</span><br><span class="line">java.io.IOException: Callable timed out after 10000 ms on file: /origin_data/compass/edb/edb_order/1970-01-01/edb.1647953566601.lzo.tmp</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter.callWithTimeout(BucketWriter.java:682)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter.renameBucket(BucketWriter.java:614)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter.access$1000(BucketWriter.java:58)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter$4.call(BucketWriter.java:330)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter$4.call(BucketWriter.java:315)</span><br><span class="line">	at java.util.concurrent.FutureTask.run(FutureTask.java:266)</span><br><span class="line">	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)</span><br><span class="line">	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)</span><br><span class="line">	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)</span><br><span class="line">	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)</span><br><span class="line">	at java.lang.Thread.run(Thread.java:748)</span><br><span class="line">Caused by: java.util.concurrent.TimeoutException</span><br><span class="line">	at java.util.concurrent.FutureTask.get(FutureTask.java:205)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter.callWithTimeout(BucketWriter.java:675)</span><br><span class="line">	... 10 more</span><br><span class="line">22 三月 2022 21:19:21,821 INFO  [PollableSourceRunner-KafkaSource-r1] (org.apache.kafka.clients.consumer.internals.AbstractCoordinator.coordinatorDead:529)  - Marking the coordinator 2147483646 dead.</span><br><span class="line">22 三月 2022 21:19:22,049 ERROR [PollableSourceRunner-KafkaSource-r1] (org.apache.kafka.clients.consumer.internals.ConsumerCoordinator$OffsetCommitResponseHandler.handle:550)  - Error UNKNOWN_MEMBER_ID occurred while committing offsets for group edb_order_consumer</span><br><span class="line">22 三月 2022 21:19:22,049 ERROR [PollableSourceRunner-KafkaSource-r1] (org.apache.flume.source.kafka.KafkaSource.doProcess:314)  - KafkaSource EXCEPTION, &#123;&#125;</span><br><span class="line">org.apache.kafka.clients.consumer.CommitFailedException: Commit cannot be completed due to group rebalance</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator$OffsetCommitResponseHandler.handle(ConsumerCoordinator.java:552)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator$OffsetCommitResponseHandler.handle(ConsumerCoordinator.java:493)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator$CoordinatorResponseHandler.onSuccess(AbstractCoordinator.java:665)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator$CoordinatorResponseHandler.onSuccess(AbstractCoordinator.java:644)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.RequestFuture$1.onSuccess(RequestFuture.java:167)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.RequestFuture.fireSuccess(RequestFuture.java:133)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.RequestFuture.complete(RequestFuture.java:107)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient$RequestFutureCompletionHandler.onComplete(ConsumerNetworkClient.java:380)</span><br><span class="line">	at org.apache.kafka.clients.NetworkClient.poll(NetworkClient.java:274)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.clientPoll(ConsumerNetworkClient.java:320)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:213)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:193)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:163)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.commitOffsetsSync(ConsumerCoordinator.java:358)</span><br><span class="line">	at org.apache.kafka.clients.consumer.KafkaConsumer.commitSync(KafkaConsumer.java:968)</span><br><span class="line">	at org.apache.flume.source.kafka.KafkaSource.doProcess(KafkaSource.java:304)</span><br><span class="line">	at org.apache.flume.source.AbstractPollableSource.process(AbstractPollableSource.java:60)</span><br><span class="line">	at org.apache.flume.source.PollableSourceRunner$PollingRunner.run(PollableSourceRunner.java:133)</span><br><span class="line">	at java.lang.Thread.run(Thread.java:748)</span><br><span class="line">22 三月 2022 21:19:22,494 INFO  [SinkRunner-PollingRunner-DefaultSinkProcessor] (org.apache.flume.sink.hdfs.HDFSCompressedDataStream.configure:64)  - Serializer = TEXT, UseRawLocalFileSystem = false</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">23 三月 2022 10:15:59,718 INFO  [agent-shutdown-hook] (org.apache.flume.instrumentation.MonitoredCounterGroup.stop:177)  - Shutdown Metric for type: SINK, name: k1. sink.event.drain.sucess == 2299667</span><br><span class="line">23 三月 2022 10:15:59,735 WARN  [agent-shutdown-hook] (org.apache.kafka.common.network.Selector.poll:309)  - Unexpected error from cos-bigdata-test-hadoop-01/192.168.101.184; closing connection</span><br><span class="line">java.lang.NullPointerException</span><br><span class="line">	at org.apache.kafka.common.network.NetworkReceive.complete(NetworkReceive.java:67)</span><br><span class="line">	at org.apache.kafka.common.network.KafkaChannel.read(KafkaChannel.java:135)</span><br><span class="line">	at org.apache.kafka.common.network.Selector.poll(Selector.java:286)</span><br><span class="line">	at org.apache.kafka.clients.NetworkClient.poll(NetworkClient.java:256)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.clientPoll(ConsumerNetworkClient.java:320)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:213)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:178)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.sendLeaveGroupRequest(AbstractCoordinator.java:575)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.maybeLeaveGroup(AbstractCoordinator.java:552)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.close(AbstractCoordinator.java:541)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.close(ConsumerCoordinator.java:321)</span><br><span class="line">	at org.apache.kafka.clients.ClientUtils.closeQuietly(ClientUtils.java:63)</span><br><span class="line">	at org.apache.kafka.clients.consumer.KafkaConsumer.close(KafkaConsumer.java:1277)</span><br><span class="line">	at org.apache.kafka.clients.consumer.KafkaConsumer.close(KafkaConsumer.java:1258)</span><br><span class="line">	at org.apache.flume.source.kafka.KafkaSource.doStop(KafkaSource.java:528)</span><br><span class="line">	at org.apache.flume.source.BasicSourceSemantics.stop(BasicSourceSemantics.java:96)</span><br><span class="line">	at org.apache.flume.source.PollableSourceRunner.stop(PollableSourceRunner.java:101)</span><br><span class="line">	at org.apache.flume.lifecycle.LifecycleSupervisor.stop(LifecycleSupervisor.java:104)</span><br><span class="line">	at org.apache.flume.node.Application.stop(Application.java:92)</span><br><span class="line">	at org.apache.flume.node.Application$1.run(Application.java:343)</span><br><span class="line">23 三月 2022 10:16:00,124 INFO  [agent-shutdown-hook] (org.apache.kafka.clients.consumer.internals.AbstractCoordinator.coordinatorDead:529)  - Marking the coordinator 2147483646 dead.</span><br></pre></td></tr></table></figure></li>
<li>原因：估计是azkaban任务运行导致的内存和cpu占用。</li>
<li>解决：</li>
</ul>
<h3 id="踩坑四十九"><a href="#踩坑四十九" class="headerlink" title="踩坑四十九"></a>踩坑四十九</h3><ul>
<li>问题：启动Flume后消费Kafka的数据一直没有反应，查看启动Flume时的日志，报错如下<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">Exception in thread &quot;PollableSourceRunner-KafkaSource-r1&quot; java.lang.OutOfMemoryError: Java heap space</span><br><span class="line">	at org.apache.kafka.common.utils.Utils.toArray(Utils.java:268)</span><br><span class="line">	at org.apache.kafka.common.utils.Utils.toArray(Utils.java:261)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.Fetcher.parseRecord(Fetcher.java:622)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.Fetcher.handleFetchResponse(Fetcher.java:566)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.Fetcher.access$000(Fetcher.java:69)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.Fetcher$1.onSuccess(Fetcher.java:139)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.Fetcher$1.onSuccess(Fetcher.java:136)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.RequestFuture.fireSuccess(RequestFuture.java:133)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.RequestFuture.complete(RequestFuture.java:107)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient$RequestFutureCompletionHandler.onComplete(ConsumerNetworkClient.java:380)</span><br><span class="line">	at org.apache.kafka.clients.NetworkClient.poll(NetworkClient.java:274)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.clientPoll(ConsumerNetworkClient.java:320)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:213)</span><br><span class="line">	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:193)</span><br><span class="line">	at org.apache.kafka.clients.consumer.KafkaConsumer.pollOnce(KafkaConsumer.java:908)</span><br><span class="line">	at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:853)</span><br><span class="line">	at org.apache.flume.source.kafka.KafkaSource.doProcess(KafkaSource.java:200)</span><br><span class="line">	at org.apache.flume.source.AbstractPollableSource.process(AbstractPollableSource.java:60)</span><br><span class="line">	at org.apache.flume.source.PollableSourceRunner$PollingRunner.run(PollableSourceRunner.java:133)</span><br><span class="line">	at java.lang.Thread.run(Thread.java:748)</span><br></pre></td></tr></table></figure></li>
<li>原因：Flume的JVM配置的内存过小</li>
<li>解决：修改conf&#x2F;flume-env.sh文件<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">export JAVA_OPTS=&quot;-Xms512m -Xmx1024m -Dcom.sun.management.jmxremote&quot;</span><br></pre></td></tr></table></figure></li>
</ul>
<h3 id="踩坑五十九"><a href="#踩坑五十九" class="headerlink" title="踩坑五十九"></a>踩坑五十九</h3><ul>
<li>问题：Flume中的hdfs sink配置了回滚条件，但是实际罗盘文件却都是小文件。在flume的日志配置文件中配置日志存储位置，查看日志后发现如下异常<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">17 六月 2022 09:43:44,663 ERROR [SinkRunner-PollingRunner-DefaultSinkProcessor] (org.apache.flume.sink.hdfs.BucketWriter.append:516)  - Hit max consecutive under-replication rotations (30); will not continue rolling files under this path due to under-replication</span><br></pre></td></tr></table></figure></li>
<li>原因：如果感知到文件正在进行块复制，那么Flume会滚动该文件并产生新的文件。而集群中有一个节点的Datanode进程掉线了(配置副本数为3且只有3个Datanode节点)，导致该路径被不同重试复制。</li>
<li>解决： 启动该掉线的节点<br>或 配置让程序感知不到写的文件所在块正在复制，在任务文件中配置<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">a1.sinks.k1.hdfs.minBlockReplicas = 1</span><br></pre></td></tr></table></figure></li>
</ul>
<h3 id="踩坑六十"><a href="#踩坑六十" class="headerlink" title="踩坑六十"></a>踩坑六十</h3><ul>
<li>问题：Flume启动kafka source任务，如果Kafka没有到达新数据，Flume消费不到任何数据，指定的消费者组没有消费者<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">Consumer group &#x27;youmeng3-active&#x27; has no active members</span><br></pre></td></tr></table></figure></li>
<li>原因：未添加消费者</li>
<li>解决：添加消费者</li>
</ul>
<h3 id="踩坑六十六"><a href="#踩坑六十六" class="headerlink" title="踩坑六十六"></a>踩坑六十六</h3><ul>
<li>问题：Flume从Kafka读取飞燕的数据落盘到hdfs中，到13:30左右时报错如下<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br></pre></td><td class="code"><pre><span class="line">23 六月 2022 15:46:22,807 ERROR [SinkRunner-PollingRunner-DefaultSinkProcessor] (org.apache.flume.SinkRunner$PollingRunner.run:158)  - Unable to deliver event. Exception follows.</span><br><span class="line">org.apache.flume.EventDeliveryException: org.apache.flume.auth.SecurityException: Privileged action failed</span><br><span class="line">	at org.apache.flume.sink.hdfs.HDFSEventSink.process(HDFSEventSink.java:451)</span><br><span class="line">	at org.apache.flume.sink.DefaultSinkProcessor.process(DefaultSinkProcessor.java:67)</span><br><span class="line">	at org.apache.flume.SinkRunner$PollingRunner.run(SinkRunner.java:145)</span><br><span class="line">	at java.lang.Thread.run(Thread.java:748)</span><br><span class="line">Caused by: org.apache.flume.auth.SecurityException: Privileged action failed</span><br><span class="line">	at org.apache.flume.auth.UGIExecutor.execute(UGIExecutor.java:49)</span><br><span class="line">	at org.apache.flume.auth.KerberosAuthenticator.execute(KerberosAuthenticator.java:64)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter$9.call(BucketWriter.java:665)</span><br><span class="line">	at java.util.concurrent.FutureTask.run(FutureTask.java:266)</span><br><span class="line">	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)</span><br><span class="line">	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)</span><br><span class="line">	... 1 more</span><br><span class="line">Caused by: java.io.IOException: write beyond end of stream</span><br><span class="line">	at com.hadoop.compression.lzo.LzopOutputStream.write(LzopOutputStream.java:134)</span><br><span class="line">	at java.io.OutputStream.write(OutputStream.java:75)</span><br><span class="line">	at org.apache.flume.serialization.BodyTextEventSerializer.write(BodyTextEventSerializer.java:70)</span><br><span class="line">	at org.apache.flume.sink.hdfs.HDFSCompressedDataStream.append(HDFSCompressedDataStream.java:125)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter$7.call(BucketWriter.java:540)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter$7.call(BucketWriter.java:537)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter$9$1.run(BucketWriter.java:668)</span><br><span class="line">	at java.security.AccessController.doPrivileged(Native Method)</span><br><span class="line">	at javax.security.auth.Subject.doAs(Subject.java:422)</span><br><span class="line">	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1729)</span><br><span class="line">	at org.apache.flume.auth.UGIExecutor.execute(UGIExecutor.java:47)</span><br><span class="line">	... 6 more</span><br><span class="line">23 六月 2022 15:46:22,805 ERROR [SinkRunner-PollingRunner-DefaultSinkProcessor] (org.apache.flume.sink.hdfs.AbstractHDFSWriter.isUnderReplicated:99)  - Unexpected error while checking replication factor</span><br><span class="line">java.lang.reflect.InvocationTargetException</span><br><span class="line">	at sun.reflect.GeneratedMethodAccessor6.invoke(Unknown Source)</span><br><span class="line">	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)</span><br><span class="line">	at java.lang.reflect.Method.invoke(Method.java:498)</span><br><span class="line">	at org.apache.flume.sink.hdfs.AbstractHDFSWriter.getNumCurrentReplicas(AbstractHDFSWriter.java:166)</span><br><span class="line">	at org.apache.flume.sink.hdfs.AbstractHDFSWriter.isUnderReplicated(AbstractHDFSWriter.java:85)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter.shouldRotate(BucketWriter.java:573)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter.append(BucketWriter.java:508)</span><br><span class="line">	at org.apache.flume.sink.hdfs.HDFSEventSink.process(HDFSEventSink.java:406)</span><br><span class="line">	at org.apache.flume.sink.DefaultSinkProcessor.process(DefaultSinkProcessor.java:67)</span><br><span class="line">	at org.apache.flume.SinkRunner$PollingRunner.run(SinkRunner.java:145)</span><br><span class="line">	at java.lang.Thread.run(Thread.java:748)</span><br><span class="line">Caused by: java.io.IOException: Failed to replace a bad datanode on the existing pipeline due to no more good datanodes being available to try. (Nodes: current=[DatanodeInfoWithStorage[192.168.101.184:9866,DS-8687386a-b9ab-4c73-84ed-9a9dac98662e,DISK], DatanodeInfoWithStorage[192.168.101.185:9866,DS-564db0fe-1d53-418f-937f-470faf52b712,DISK]], original=[DatanodeInfoWithStorage[192.168.101.184:9866,DS-8687386a-b9ab-4c73-84ed-9a9dac98662e,DISK], DatanodeInfoWithStorage[192.168.101.185:9866,DS-564db0fe-1d53-418f-937f-470faf52b712,DISK]]). The current failed datanode replacement policy is DEFAULT, and a client may configure this via &#x27;dfs.client.block.write.replace-datanode-on-failure.policy&#x27; in its configuration.</span><br><span class="line">	at org.apache.hadoop.hdfs.DataStreamer.findNewDatanode(DataStreamer.java:1304)</span><br><span class="line">	at org.apache.hadoop.hdfs.DataStreamer.addDatanode2ExistingPipeline(DataStreamer.java:1372)</span><br><span class="line">	at org.apache.hadoop.hdfs.DataStreamer.handleDatanodeReplacement(DataStreamer.java:1598)</span><br><span class="line">	at org.apache.hadoop.hdfs.DataStreamer.setupPipelineInternal(DataStreamer.java:1499)</span><br><span class="line">	at org.apache.hadoop.hdfs.DataStreamer.setupPipelineForAppendOrRecovery(DataStreamer.java:1481)</span><br><span class="line">	at org.apache.hadoop.hdfs.DataStreamer.processDatanodeOrExternalError(DataStreamer.java:1256)</span><br><span class="line">	at org.apache.hadoop.hdfs.DataStreamer.run(DataStreamer.java:667)</span><br><span class="line">23 六月 2022 15:46:22,808 ERROR [SinkRunner-PollingRunner-DefaultSinkProcessor] (org.apache.flume.sink.hdfs.HDFSEventSink.process:447)  - process failed</span><br><span class="line">org.apache.flume.auth.SecurityException: Privileged action failed</span><br><span class="line">	at org.apache.flume.auth.UGIExecutor.execute(UGIExecutor.java:49)</span><br><span class="line">	at org.apache.flume.auth.KerberosAuthenticator.execute(KerberosAuthenticator.java:64)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter$9.call(BucketWriter.java:665)</span><br><span class="line">	at java.util.concurrent.FutureTask.run(FutureTask.java:266)</span><br><span class="line">	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)</span><br><span class="line">	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)</span><br><span class="line">	at java.lang.Thread.run(Thread.java:748)</span><br><span class="line">Caused by: java.io.IOException: write beyond end of stream</span><br><span class="line">	at com.hadoop.compression.lzo.LzopOutputStream.write(LzopOutputStream.java:134)</span><br><span class="line">	at java.io.OutputStream.write(OutputStream.java:75)</span><br><span class="line">	at org.apache.flume.serialization.BodyTextEventSerializer.write(BodyTextEventSerializer.java:70)</span><br><span class="line">	at org.apache.flume.sink.hdfs.HDFSCompressedDataStream.append(HDFSCompressedDataStream.java:125)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter$7.call(BucketWriter.java:540)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter$7.call(BucketWriter.java:537)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter$9$1.run(BucketWriter.java:668)</span><br><span class="line">	at java.security.AccessController.doPrivileged(Native Method)</span><br><span class="line">	at javax.security.auth.Subject.doAs(Subject.java:422)</span><br><span class="line">	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1729)</span><br><span class="line">	at org.apache.flume.auth.UGIExecutor.execute(UGIExecutor.java:47)</span><br><span class="line">	... 6 more</span><br></pre></td></tr></table></figure></li>
<li>原因：这里涉及到hdfs的两个配置<ul>
<li><code>dfs.client.block.write.replace-datanode-on-failure.enable</code>：如果在写入管道中存在一个DataNode或者网络故障时，那么DFSClient将尝试从管道中删除失败的DataNode，然后继续尝试剩下的DataNodes进行写入。结果，管道中的DataNodes的数量在减少。(enable ：启用特性，disable：禁用特性) 该特性是在管道中添加新的DataNodes。当集群规模非常小时，例如3个节点或更少时，集群管理员可能希望将策略在默认配置文件里面设置为NEVER或者禁用该特性。否则，因为找不到新的DataNode来替换，用户可能会经历异常高的管道写入错误,导致追加文件操作失败。</li>
<li><code>dfs.client.block.write.replace-datanode-on-failure.policy</code>：这个属性只有在dfs.client.block.write.replace-datanode-on-failure.enable设置true时有效：(<strong>ALWAYS</strong>：当一个存在的DataNode被删除时，总是添加一个新的DataNode; <strong>NEVER</strong>  ：永远不添加新的DataNode; <strong>DEFAULT</strong>：副本数是r，DataNode的数时n，只要r &gt;&#x3D; 3时，或者floor(r&#x2F;2)大于等于n时，或 r&gt;n时再添加一个新的DataNode，并且这个块是hflushed&#x2F;appended)</li>
</ul>
</li>
<li>解决：设置<code>dfs.client.block.write.replace-datanode-on-failure.policy=NEVER</code></li>
</ul>
<h3 id="踩坑六十七"><a href="#踩坑六十七" class="headerlink" title="踩坑六十七"></a>踩坑六十七</h3><ul>
<li>问题：Flume启动报错如下<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">org.apache.flume.ChannelException: Cannot commit transaction. Byte capacity allocated to store event body 640000.0reached. Please increase heap space/byte capacity allocated to the channel as the sinks may not be keeping up with the sources</span><br><span class="line">	at org.apache.flume.channel.MemoryChannel$MemoryTransaction.doCommit(MemoryChannel.java:120)</span><br><span class="line">	at org.apache.flume.channel.BasicTransactionSemantics.commit(BasicTransactionSemantics.java:151)</span><br><span class="line">	at org.apache.flume.channel.ChannelProcessor.processEventBatch(ChannelProcessor.java:194)</span><br><span class="line">	at org.apache.flume.source.kafka.KafkaSource.doProcess(KafkaSource.java:295)</span><br><span class="line">	at org.apache.flume.source.AbstractPollableSource.process(AbstractPollableSource.java:60)</span><br><span class="line">	at org.apache.flume.source.PollableSourceRunner$PollingRunner.run(PollableSourceRunner.java:133)</span><br><span class="line">	at java.lang.Thread.run(Thread.java:748)</span><br></pre></td></tr></table></figure></li>
<li>原因： source 接收端的容量达到上线，无法再接收数据</li>
<li>解决：</li>
</ul>
<h3 id="踩坑六十八"><a href="#踩坑六十八" class="headerlink" title="踩坑六十八"></a>踩坑六十八</h3><ul>
<li>问题：Flume在运行中报错如下<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">23 六月 2022 19:02:49,798 ERROR [SinkRunner-PollingRunner-DefaultSinkProcessor] (org.apache.flume.sink.hdfs.HDFSEventSink.process:447)  - process failed</span><br><span class="line">java.lang.InterruptedException: Timed out before HDFS call was made. Your hdfs.callTimeout might be set too low or HDFS calls are taking too long.</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter.checkAndThrowInterruptedException(BucketWriter.java:649)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter.flush(BucketWriter.java:409)</span><br><span class="line">	at org.apache.flume.sink.hdfs.HDFSEventSink.process(HDFSEventSink.java:430)</span><br><span class="line">	at org.apache.flume.sink.DefaultSinkProcessor.process(DefaultSinkProcessor.java:67)</span><br><span class="line">	at org.apache.flume.SinkRunner$PollingRunner.run(SinkRunner.java:145)</span><br><span class="line">	at java.lang.Thread.run(Thread.java:748)</span><br><span class="line">23 六月 2022 19:02:49,799 ERROR [SinkRunner-PollingRunner-DefaultSinkProcessor] (org.apache.flume.SinkRunner$PollingRunner.run:158)  - Unable to deliver event. Exception follows.</span><br><span class="line">org.apache.flume.EventDeliveryException: java.lang.InterruptedException: Timed out before HDFS call was made. Your hdfs.callTimeout might be set too low or HDFS calls are taking too long.</span><br><span class="line">	at org.apache.flume.sink.hdfs.HDFSEventSink.process(HDFSEventSink.java:451)</span><br><span class="line">	at org.apache.flume.sink.DefaultSinkProcessor.process(DefaultSinkProcessor.java:67)</span><br><span class="line">	at org.apache.flume.SinkRunner$PollingRunner.run(SinkRunner.java:145)</span><br><span class="line">	at java.lang.Thread.run(Thread.java:748)</span><br><span class="line">Caused by: java.lang.InterruptedException: Timed out before HDFS call was made. Your hdfs.callTimeout might be set too low or HDFS calls are taking too long.</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter.checkAndThrowInterruptedException(BucketWriter.java:649)</span><br><span class="line">	at org.apache.flume.sink.hdfs.BucketWriter.flush(BucketWriter.java:409)</span><br><span class="line">	at org.apache.flume.sink.hdfs.HDFSEventSink.process(HDFSEventSink.java:430)</span><br><span class="line">	... 3 more</span><br></pre></td></tr></table></figure></li>
<li>原因：</li>
<li>解决：</li>
</ul>
<h3 id="踩坑六十九"><a href="#踩坑六十九" class="headerlink" title="踩坑六十九"></a>踩坑六十九</h3><ul>
<li>现象：Datax从hdfs往mysql导数据，但是卡在加载驱动这一步不动了<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">28-06-2022 09:22:03 CST cb_sales2mysql INFO - 2022-06-28 09:22:03.653 [main] WARN  Engine - prioriy set to 0, because NumberFormatException, the value is: null</span><br><span class="line">28-06-2022 09:22:03 CST cb_sales2mysql INFO - 2022-06-28 09:22:03.655 [main] INFO  PerfTrace - PerfTrace traceId=job_-1, isEnable=false, priority=0</span><br><span class="line">28-06-2022 09:22:03 CST cb_sales2mysql INFO - 2022-06-28 09:22:03.655 [main] INFO  JobContainer - DataX jobContainer starts job.</span><br><span class="line">28-06-2022 09:22:03 CST cb_sales2mysql INFO - 2022-06-28 09:22:03.657 [main] INFO  JobContainer - Set jobId = 0</span><br><span class="line">28-06-2022 09:22:03 CST cb_sales2mysql INFO - 2022-06-28 09:22:03.669 [job-0] INFO  HdfsReader$Job - init() begin...</span><br><span class="line">28-06-2022 09:22:03 CST cb_sales2mysql INFO - 2022-06-28 09:22:03.927 [job-0] INFO  HdfsReader$Job - hadoopConfig details:&#123;&quot;finalParameters&quot;:[]&#125;</span><br><span class="line">28-06-2022 09:22:03 CST cb_sales2mysql INFO - 2022-06-28 09:22:03.927 [job-0] INFO  HdfsReader$Job - init() ok and end...</span><br><span class="line">28-06-2022 09:22:03 CST cb_sales2mysql INFO - Loading class `com.mysql.jdbc.Driver&#x27;. This is deprecated. The new driver class is `com.mysql.cj.jdbc.Driver&#x27;. The driver is automatically registered via the SPI and manual loading of the driver class is generally unnecessary.</span><br></pre></td></tr></table></figure></li>
<li>原因：mysql表中出现了死锁，导致整张表锁死，无法写入数据</li>
<li>解决：解决mysql表的死锁问题</li>
</ul>
<h3 id="踩坑七十"><a href="#踩坑七十" class="headerlink" title="踩坑七十"></a>踩坑七十</h3><ul>
<li>现象：使用Fair Scheduler调度Tez任务时，经常发生异常如下<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br></pre></td><td class="code"><pre><span class="line">2022-06-29 16:32:37,779 [FATAL] [HistoryEventHandlingThread] |yarn.YarnUncaughtExceptionHandler|: Thread Thread[HistoryEventHandlingThread,5,main] threw an Error.  Shutting down now...</span><br><span class="line">java.lang.VerifyError: Stack map does not match the one at exception handler 77</span><br><span class="line">Exception Details:</span><br><span class="line">  Location:</span><br><span class="line">    com/fasterxml/jackson/databind/deser/std/StdDeserializer._parseDate(Lcom/fasterxml/jackson/core/JsonParser;Lcom/fasterxml/jackson/databind/DeserializationContext;)Ljava/util/Date; @77: astore</span><br><span class="line">  Reason:</span><br><span class="line">    Type &#x27;com/fasterxml/jackson/core/JsonParseException&#x27; (current frame, stack[0]) is not assignable to &#x27;com/fasterxml/jackson/core/exc/StreamReadException&#x27; (stack map, stack[0])</span><br><span class="line">  Current Frame:</span><br><span class="line">    bci: @69</span><br><span class="line">    flags: &#123; &#125;</span><br><span class="line">    locals: &#123; &#x27;com/fasterxml/jackson/databind/deser/std/StdDeserializer&#x27;, &#x27;com/fasterxml/jackson/core/JsonParser&#x27;, &#x27;com/fasterxml/jackson/databind/DeserializationContext&#x27; &#125;</span><br><span class="line">    stack: &#123; &#x27;com/fasterxml/jackson/core/JsonParseException&#x27; &#125;</span><br><span class="line">  Stackmap Frame:</span><br><span class="line">    bci: @77</span><br><span class="line">    flags: &#123; &#125;</span><br><span class="line">    locals: &#123; &#x27;com/fasterxml/jackson/databind/deser/std/StdDeserializer&#x27;, &#x27;com/fasterxml/jackson/core/JsonParser&#x27;, &#x27;com/fasterxml/jackson/databind/DeserializationContext&#x27; &#125;</span><br><span class="line">    stack: &#123; &#x27;com/fasterxml/jackson/core/exc/StreamReadException&#x27; &#125;</span><br><span class="line">  Bytecode:</span><br><span class="line">    0x0000000: 2bb6 0035 aa00 0000 0000 0081 0000 0003</span><br><span class="line">    0x0000010: 0000 000b 0000 007a 0000 0081 0000 0081</span><br><span class="line">    0x0000020: 0000 0034 0000 0041 0000 0081 0000 0081</span><br><span class="line">    0x0000030: 0000 0081 0000 0071 2a2b b600 11b6 0012</span><br><span class="line">    0x0000040: 2cb6 006b b02b b600 4742 a700 223a 052c</span><br><span class="line">    0x0000050: 2ab4 0002 2bb6 006e 126f 03bd 0004 b600</span><br><span class="line">    0x0000060: 70c0 002d 3a06 1906 b600 4c42 bb00 7159</span><br><span class="line">    0x0000070: 21b7 0072 b02a 2cb6 0073 c000 71b0 2a2b</span><br><span class="line">    0x0000080: 2cb6 0074 b02c 2ab4 0002 2bb6 0025 c000</span><br><span class="line">    0x0000090: 71b0                                   </span><br><span class="line">  Exception Handler Table:</span><br><span class="line">    bci [69, 74] =&gt; handler: 77</span><br><span class="line">    bci [69, 74] =&gt; handler: 77</span><br><span class="line">  Stackmap Table:</span><br><span class="line">    same_frame(@56)</span><br><span class="line">    same_frame(@69)</span><br><span class="line">    same_locals_1_stack_item_frame(@77,Object[#359])</span><br><span class="line">    append_frame(@108,Long)</span><br><span class="line">    chop_frame(@117,1)</span><br><span class="line">    same_frame(@126)</span><br><span class="line">    same_frame(@133)</span><br><span class="line"></span><br><span class="line">	at com.fasterxml.jackson.databind.deser.std.JdkDeserializers.&lt;clinit&gt;(JdkDeserializers.java:26)</span><br><span class="line">	at com.fasterxml.jackson.databind.deser.BasicDeserializerFactory.findDefaultDeserializer(BasicDeserializerFactory.java:1852)</span><br><span class="line">	at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.findStdDeserializer(BeanDeserializerFactory.java:167)</span><br><span class="line">	at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.createBeanDeserializer(BeanDeserializerFactory.java:131)</span><br><span class="line">	at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer2(DeserializerCache.java:411)</span><br><span class="line">	at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer(DeserializerCache.java:349)</span><br><span class="line">	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:264)</span><br><span class="line">	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer(DeserializerCache.java:244)</span><br><span class="line">	at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:142)</span><br><span class="line">	at com.fasterxml.jackson.databind.DeserializationContext.findRootValueDeserializer(DeserializationContext.java:476)</span><br><span class="line">	at com.fasterxml.jackson.databind.ObjectReader._prefetchRootDeserializer(ObjectReader.java:2094)</span><br><span class="line">	at com.fasterxml.jackson.databind.ObjectReader.forType(ObjectReader.java:681)</span><br><span class="line">	at com.fasterxml.jackson.jaxrs.base.ProviderBase.readFrom(ProviderBase.java:799)</span><br><span class="line">	at com.sun.jersey.api.client.ClientResponse.getEntity(ClientResponse.java:634)</span><br><span class="line">	at com.sun.jersey.api.client.ClientResponse.getEntity(ClientResponse.java:586)</span><br><span class="line">	at org.apache.hadoop.yarn.client.api.impl.TimelineWriter.putEntities(TimelineWriter.java:93)</span><br><span class="line">	at org.apache.hadoop.yarn.client.api.impl.TimelineClientImpl.putEntities(TimelineClientImpl.java:177)</span><br><span class="line">	at org.apache.tez.dag.history.logging.ats.ATSHistoryLoggingService.handleEvents(ATSHistoryLoggingService.java:354)</span><br><span class="line">	at org.apache.tez.dag.history.logging.ats.ATSHistoryLoggingService.access$600(ATSHistoryLoggingService.java:59)</span><br><span class="line">	at org.apache.tez.dag.history.logging.ats.ATSHistoryLoggingService$1.run(ATSHistoryLoggingService.java:206)</span><br><span class="line">	at java.lang.Thread.run(Thread.java:748)</span><br></pre></td></tr></table></figure></li>
<li>原因：java.lang.VerifyError 是说 JVM 在加载一个类时，会去校验类的正确性，只有类文件不合法才会报这个Error。可能是tez中的jackson包的版本(2.10.0)与hadoop中的jackson包版本(2.7.8)不一致导致的。在2.10.0版本中存在com&#x2F;fasterxml&#x2F;jackson&#x2F;core&#x2F;exc&#x2F;StreamReadException类，但是2.7.8中不存在，找不到类会导致编译不通过。<br> 而集群中只有179节点的tez中添加了jackson-core-2.10.0，所以azkaban分配到该节点就会发生这个错误。</li>
<li>解决：替换tez中的jackson包的版本与hadoop中的一致(hadoop3.1.3的jackson版本为2.7.8)。</li>
<li>参考：<a target="_blank" rel="noopener" href="https://yrom.net/blog/2016/08/22/java-lang-verifyerror-on-android/">Android 不想和你说话，抛了个 java.lang.VerifyError - Yrom’s</a></li>
</ul>
<h3 id="踩坑七十一"><a href="#踩坑七十一" class="headerlink" title="踩坑七十一"></a>踩坑七十一</h3><ul>
<li>现象：Kafka的topic有两个副本，但是缺失了一个副本同步。<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">[root@cos-bigdata-hadoop-01 kafka-2.11]# bin/kafka-topics.sh --zookeeper 192.168.101.179:2181 --topic status-model --describe</span><br><span class="line">Topic:status-model	PartitionCount:3	ReplicationFactor:2	Configs:</span><br><span class="line">	Topic: status-model	Partition: 0	Leader: 0	Replicas: 0,2	Isr: 2,0</span><br><span class="line">	Topic: status-model	Partition: 1	Leader: 1	Replicas: 1,0	Isr: 0,1</span><br><span class="line">	Topic: status-model	Partition: 2	Leader: 2	Replicas: 2,1	Isr: 2</span><br></pre></td></tr></table></figure></li>
<li>原因：ISR丢失(抖动)</li>
<li>解决：</li>
</ul>
<h3 id="踩坑七十二"><a href="#踩坑七十二" class="headerlink" title="踩坑七十二"></a>踩坑七十二</h3><ul>
<li>现象：使用其他集群的kafka客户端来访问本集群的topic。如果该topic只有一个分区，那么可以成功访问，如果该topic有多个分区，那么就会报错如下<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">[root@cos-bigdata-test-hadoop-02 kafka_2.11]# bin/kafka-console-consumer.sh --bootstrap-server 192.168.101.181:9092 --topic compass-edb-sales-prod --from-beginning</span><br><span class="line">[2022-08-16 11:32:12,460] WARN Received unknown topic or partition error in ListOffset request for partition compass-edb-sales-prod-1. The topic/partition may not exist or the user may not have Describe access to it. (org.apache.kafka.clients.consumer.internals.Fetcher)</span><br><span class="line">[2022-08-16 11:32:12,462] WARN Received unknown topic or partition error in ListOffset request for partition compass-edb-sales-prod-0. The topic/partition may not exist or the user may not have Describe access to it. (org.apache.kafka.clients.consumer.internals.Fetcher)</span><br><span class="line">[2022-08-16 11:32:12,462] WARN Received unknown topic or partition error in ListOffset request for partition compass-edb-sales-prod-2. The topic/partition may not exist or the user may not have Describe access to it. (org.apache.kafka.clients.consumer.internals.Fetcher)</span><br></pre></td></tr></table></figure></li>
<li>原因：猜测是读取使用的命令中指定的是Kafka集群，而没有指定zookeeper，所以找不到topic对应的多个分区所在的节点。在Zookeeper的节点&#x2F;borkers&#x2F;topics&#x2F;[topic_name]&#x2F;partitions&#x2F;0&#x2F;state中记录了谁是leader，有哪些服务器可用；![image.png](集群问题.assets54318a17d494c2c94ffef958765315a.png)</li>
<li>解决：</li>
</ul>
<h3 id="踩坑七十三"><a href="#踩坑七十三" class="headerlink" title="踩坑七十三"></a>踩坑七十三</h3><ul>
<li>设置spark.executor.instances配置时，报错<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">[2022-09-17 09:59:04] Error while processing statement: Cannot modify spark.executor.cores at runtime. It is not in list of params that are allowed to be modified at runtime</span><br></pre></td></tr></table></figure></li>
<li>原因：</li>
<li>解决：hive-site.conf添加如下<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">&lt;property&gt;</span><br><span class="line">   &lt;name&gt;hive.security.authorization.sqlstd.confwhitelist&lt;/name&gt;        &lt;value&gt;mapred.*|hive.*|mapreduce.*|spark.*|tez.*&lt;/value&gt;</span><br><span class="line">&lt;/property&gt;</span><br><span class="line">&lt;property&gt;</span><br><span class="line">    &lt;name&gt;hive.security.authorization.sqlstd.confwhitelist.append&lt;/name&gt;</span><br><span class="line">    &lt;value&gt;mapred.*|hive.*|mapreduce.*|spark.*|tez.*&lt;/value&gt;</span><br><span class="line">&lt;/property&gt;</span><br></pre></td></tr></table></figure></li>
</ul>
<h3 id="踩坑七十四"><a href="#踩坑七十四" class="headerlink" title="踩坑七十四"></a>踩坑七十四</h3><ul>
<li>问题：namenode高可用配置后，运行了6个小时，active namenode转为standby，但是没有选举出新的active namenode节点<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">2023-04-21 09:15:34,484 INFO org.apache.hadoop.hdfs.server.namenode.ha.EditLogTailer: Triggering log roll on remote NameNode</span><br><span class="line">2023-04-21 09:15:34,493 WARN org.apache.hadoop.hdfs.server.namenode.ha.EditLogTailer: Exception from remote name node RemoteNameNodeInfo [nnId=nn2, ipcAddress=cos-bigdata-test-flink-02/192.168.101.194:8020, httpAddress=http://cos-bigdata-test-flink-02:9870], try next.</span><br><span class="line">org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.ipc.StandbyException): Operation category JOURNAL is not supported in state standby. Visit https://s.apache.org/sbnn-error</span><br><span class="line">2023-04-21 09:15:34,497 WARN org.apache.hadoop.hdfs.server.namenode.ha.EditLogTailer: Exception from remote name node RemoteNameNodeInfo [nnId=nn3, ipcAddress=cos-bigdata-test-flink-03/192.168.101.195:8020, httpAddress=http://cos-bigdata-test-flink-03:9870], try next.</span><br><span class="line">org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.ipc.StandbyException): Operation category JOURNAL is not supported in state standby. Visit https://s.apache.org/sbnn-error</span><br><span class="line">2023-04-21 09:15:34,504 WARN org.apache.hadoop.hdfs.server.namenode.ha.EditLogTailer: Unable to trigger a roll of the active NN</span><br><span class="line">java.util.concurrent.ExecutionException: java.io.IOException: Cannot find any valid remote NN to service request!</span><br><span class="line">	at java.util.concurrent.FutureTask.report(FutureTask.java:122)</span><br><span class="line">	at java.util.concurrent.FutureTask.get(FutureTask.java:206)</span><br><span class="line">	at org.apache.hadoop.hdfs.server.namenode.ha.EditLogTailer.triggerActiveLogRoll(EditLogTailer.java:416)</span><br><span class="line">	at org.apache.hadoop.hdfs.server.namenode.ha.EditLogTailer$EditLogTailerThread.doWork(EditLogTailer.java:475)</span><br><span class="line">	at org.apache.hadoop.hdfs.server.namenode.ha.EditLogTailer$EditLogTailerThread.access$300(EditLogTailer.java:441)</span><br><span class="line">	at org.apache.hadoop.hdfs.server.namenode.ha.EditLogTailer$EditLogTailerThread$1.run(EditLogTailer.java:458)</span><br><span class="line">	at org.apache.hadoop.security.SecurityUtil.doAsLoginUserOrFatal(SecurityUtil.java:482)</span><br><span class="line">	at org.apache.hadoop.hdfs.server.namenode.ha.EditLogTailer$EditLogTailerThread.run(EditLogTailer.java:454)</span><br><span class="line">Caused by: java.io.IOException: Cannot find any valid remote NN to service request!</span><br><span class="line">	at org.apache.hadoop.hdfs.server.namenode.ha.EditLogTailer$MultipleNameNodeProxy.call(EditLogTailer.java:573)</span><br><span class="line">	at java.util.concurrent.FutureTask.run(FutureTask.java:266)</span><br><span class="line">	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)</span><br><span class="line">	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)</span><br><span class="line">	at java.lang.Thread.run(Thread.java:748)</span><br></pre></td></tr></table></figure></li>
<li>原因：未找到原因</li>
<li>解决：关闭hdfs集群<code>stop-dfs.sh</code>，格式化zookeeper <code>hdfs zkfc -formatZK</code>，重启hdfs集群 <code>start-dfs.sh</code>后正常。此时再kill掉namenode后可以正常进行故障转移。</li>
</ul>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">文章作者: </span><span class="post-copyright-info"><a href="http://example.com">CJ</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">文章链接: </span><span class="post-copyright-info"><a href="http://example.com/2023/05/06/%E5%A4%A7%E6%95%B0%E6%8D%AE%E7%A6%BB%E7%BA%BF/%E9%9B%86%E7%BE%A4%E9%97%AE%E9%A2%98/">http://example.com/2023/05/06/%E5%A4%A7%E6%95%B0%E6%8D%AE%E7%A6%BB%E7%BA%BF/%E9%9B%86%E7%BE%A4%E9%97%AE%E9%A2%98/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">版权声明: </span><span class="post-copyright-info">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0</a> 许可协议。转载请注明来自 <a href="http://example.com" target="_blank">Hexo</a>！</span></div></div><div class="tag_share"><div class="post-meta__tag-list"></div><div class="post_share"><div class="social-share" data-image="https://i.loli.net/2021/02/24/5O1day2nriDzjSu.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2023/05/06/%E5%A4%A7%E6%95%B0%E6%8D%AE%E7%A6%BB%E7%BA%BF/Zookeeper-3-5-7/" title="Zookeeper-3-5-7"><div class="cover" style="background: var(--default-bg-color)"></div><div class="pagination-info"><div class="label">上一篇</div><div class="prev_info">Zookeeper-3-5-7</div></div></a></div><div class="next-post pull-right"><a href="/2023/05/06/%E5%A4%A7%E6%95%B0%E6%8D%AE%E7%AC%94%E8%AE%B0/Flink%E5%AE%9E%E6%97%B6%E9%A1%B9%E7%9B%AE/" title="Flink实时项目"><div class="cover" style="background: var(--default-bg-color)"></div><div class="pagination-info"><div class="label">下一篇</div><div class="next_info">Flink实时项目</div></div></a></div></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src="https://i.loli.net/2021/02/24/5O1day2nriDzjSu.png" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info__name">CJ</div><div class="author-info__description"></div></div><div class="card-info-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">419</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">0</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">38</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/xxxxxx"><i class="fab fa-github"></i><span>Follow Me</span></a></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>公告</span></div><div class="announcement_content">This is my Blog</div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>目录</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E9%97%AE%E9%A2%98%E4%B8%80"><span class="toc-number">1.</span> <span class="toc-text">问题一</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E9%97%AE%E9%A2%98%E4%BA%8C"><span class="toc-number">2.</span> <span class="toc-text">问题二</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E9%97%AE%E9%A2%98%E4%B8%89"><span class="toc-number">3.</span> <span class="toc-text">问题三</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E9%97%AE%E9%A2%98%E5%9B%9B"><span class="toc-number">4.</span> <span class="toc-text">问题四</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%B8%A9%E5%9D%91%E5%8D%81%E5%85%AD"><span class="toc-number">5.</span> <span class="toc-text">踩坑十六</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%B8%A9%E5%9D%91%E5%9B%9B%E5%8D%81%E5%85%AB"><span class="toc-number">6.</span> <span class="toc-text">踩坑四十八</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%B8%A9%E5%9D%91%E5%9B%9B%E5%8D%81%E4%B9%9D"><span class="toc-number">7.</span> <span class="toc-text">踩坑四十九</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%B8%A9%E5%9D%91%E4%BA%94%E5%8D%81%E4%B9%9D"><span class="toc-number">8.</span> <span class="toc-text">踩坑五十九</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%B8%A9%E5%9D%91%E5%85%AD%E5%8D%81"><span class="toc-number">9.</span> <span class="toc-text">踩坑六十</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%B8%A9%E5%9D%91%E5%85%AD%E5%8D%81%E5%85%AD"><span class="toc-number">10.</span> <span class="toc-text">踩坑六十六</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%B8%A9%E5%9D%91%E5%85%AD%E5%8D%81%E4%B8%83"><span class="toc-number">11.</span> <span class="toc-text">踩坑六十七</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%B8%A9%E5%9D%91%E5%85%AD%E5%8D%81%E5%85%AB"><span class="toc-number">12.</span> <span class="toc-text">踩坑六十八</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%B8%A9%E5%9D%91%E5%85%AD%E5%8D%81%E4%B9%9D"><span class="toc-number">13.</span> <span class="toc-text">踩坑六十九</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%B8%A9%E5%9D%91%E4%B8%83%E5%8D%81"><span class="toc-number">14.</span> <span class="toc-text">踩坑七十</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%B8%A9%E5%9D%91%E4%B8%83%E5%8D%81%E4%B8%80"><span class="toc-number">15.</span> <span class="toc-text">踩坑七十一</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%B8%A9%E5%9D%91%E4%B8%83%E5%8D%81%E4%BA%8C"><span class="toc-number">16.</span> <span class="toc-text">踩坑七十二</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%B8%A9%E5%9D%91%E4%B8%83%E5%8D%81%E4%B8%89"><span class="toc-number">17.</span> <span class="toc-text">踩坑七十三</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%B8%A9%E5%9D%91%E4%B8%83%E5%8D%81%E5%9B%9B"><span class="toc-number">18.</span> <span class="toc-text">踩坑七十四</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>最新文章</span></div><div class="aside-list"><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2023/05/06/MySQL/%E6%B3%A8%E8%A7%A3@Select%E5%92%8C@Insert/" title="注解@Select和@Insert">注解@Select和@Insert</a><time datetime="2023-05-06T05:48:28.906Z" title="发表于 2023-05-06 13:48:28">2023-05-06</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2023/05/06/%E5%90%8E%E7%AB%AF%E6%A1%86%E6%9E%B6/%E6%B3%A8%E8%A7%A3@EnableAutoConfiguration/" title="注解@EnableAutoConfiguration">注解@EnableAutoConfiguration</a><time datetime="2023-05-06T05:48:06.027Z" title="发表于 2023-05-06 13:48:06">2023-05-06</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2023/05/06/%E5%A4%A7%E6%95%B0%E6%8D%AE%E7%A6%BB%E7%BA%BF/%E5%A4%A7%E6%95%B0%E6%8D%AE%E9%9B%86%E7%BE%A4%E7%9B%91%E6%8E%A7%E6%A1%86%E6%9E%B6/" title="大数据集群监控框架">大数据集群监控框架</a><time datetime="2023-05-06T05:42:56.298Z" title="发表于 2023-05-06 13:42:56">2023-05-06</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2023/05/06/%E9%AB%98%E5%B9%B6%E5%8F%91/HashMap%E5%B9%B6%E5%8F%91%E9%97%AE%E9%A2%98%E5%8F%8AConcurrentHashMap%E5%8E%9F%E7%90%86/" title="HashMap并发问题及ConcurrentHashMap原理">HashMap并发问题及ConcurrentHashMap原理</a><time datetime="2023-05-06T05:31:21.103Z" title="发表于 2023-05-06 13:31:21">2023-05-06</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2023/05/06/%E9%AB%98%E5%B9%B6%E5%8F%91/Stream%E5%8E%9F%E7%90%86/" title="Stream原理">Stream原理</a><time datetime="2023-05-06T05:31:21.103Z" title="发表于 2023-05-06 13:31:21">2023-05-06</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2020 - 2023 By CJ</div><div class="framework-info"><span>框架 </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>主题 </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="阅读模式"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="浅色和深色模式转换"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="单栏和双栏切换"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="设置"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="目录"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="回到顶部"><span class="scroll-percent"></span><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox/fancybox.umd.min.js"></script><div class="js-pjax"></div><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>